### PR TITLE
Add method to get L1 message hash

### LINKED ISSUE #1139: Add provider function to get L1 message hash (where applicable)
### Feature Request

**Describe the solution you'd like**
A provider function `getL1MessageHash` which receives an L2 tx hash, and returns the L1 message hash, if it is applicable.

**Context**
L1 message hashes are computed in the following way [link to docs](https://docs.starknet.io/documentation/architecture_and_concepts/L1-L2_Communication/messaging-mechanism/#structure_and_hashing_l1-l2).

A snippet of solidity code to compute this:
```
pragma solidity >=0.8.2 <0.9.0;
contract Hasher {
  function getL1ToL2MsgHash(
        address fromAddress,
        uint256 toAddress,
        uint256 selector,
        uint256[] calldata payload,
        uint256 nonce
    ) external view returns (bytes32) {
        return
            keccak256(
                abi.encodePacked(
                    uint256(uint160(fromAddress)),
                    toAddress,
                    nonce,
                    selector,
                    payload.length,
                    payload
                )
            );
    }
}
```

The required information to compute this on L2 is available when querying `getTransactionByHash`
Example:
mainnet tx: [0x0075f4ad64b9eca93b4e1bb48fa3db4260e476fcf65e7ab97bf5f60a2a3022ce](https://voyager.online/tx/0x7fb25547e3be9aa8db2e97cc830b4b5d487110a4e42e65b45924b9a4e2fe5f7)

The result of `getTransactionbyHash`
```
{
  "jsonrpc": "2.0",
  "id": 0,
  "result": {
    "calldata": [
      "0xae0ee0a63a2ce6baeeffe56e7714fb4efe48d419",
      "0x1d8ec38501c2fddf0b17fcd38df2029fb059b2016889e6962e5f19449735108",
      "0x354a6ba7a180000",
      "0x0"
    ],
    "contract_address": "0x73314940630fd6dcda0d772d4c972c4e0a9946bef9dabf4ef84eda8ef542b82",
    "entry_point_selector": "0x2d757788a8d8d6f21d1cd40bce38a8222d70654214e96ff95d8086e684fbee5",
    "nonce": "0x15b1a",
    "transaction_hash": "0x75f4ad64b9eca93b4e1bb48fa3db4260e476fcf65e7ab97bf5f60a2a3022ce",
    "type": "L1_HANDLER",
    "version": "0x0"
  }
}
```

The first param of `calldata` == `fromAddress`.
`contract_address` == `toAddress`
`nonce` == `nonce`
`entry_point_selector` == `selector`

`payload` and `payload.length` can be derived from the calldata field.

In case the L2 tx is not a result of an l1->l2 interaction, the function can return null (or error)

This can be a good first issue

### PR DESCRIPTION
**Pull request by ddoktorski**
Closes #1139 

## Introduced changes

- Function to get L1 message hash for `L1_Handler` transactions

Some tests from `tests_ci_on_network` expected to fail at this point. It should be fixed once #1201 is merged.

##

- [ ] This PR contains breaking changes

### CODE REVIEW

File: starknet_py/net/client_utils.py [hunk 1]
```diff
@@ -17,3 +18,21 @@ def hash_to_felt(value: Hash) -> str:
 
 def is_block_identifier(value: Union[int, Hash, Tag]) -> bool:
     return isinstance(value, str) and value in get_args(Tag)
+
+
+def encode_l1_message(tx: L1HandlerTransaction) -> bytes:
+    # TODO (#1047): remove this assert once GatewayClient is deprecated and nonce is always required
+    assert tx.nonce
```
Comment (tomek0123456789) [inline starknet_py/net/client_utils.py:25]:
```suggestion
    assert tx.nonce is not None
```

Explicit `None` checks are preferred in starknet.py (at least I was taught that 😄)

**Review by tomek0123456789 — COMMENTED** [review]
Looks good, just needs adding tests for the `utils` functions!

**Comment by ddoktorski** [comment]
Tests for `keccak256` function taken from Solidity repository:
- https://github.com/ethereum/solidity/blob/develop/test/libsolidity/semanticTests/builtinFunctions/keccak256.sol
- https://github.com/ethereum/solidity/blob/develop/test/libsolutil/Keccak256.cpp

**Review by tomek0123456789 — APPROVED** [review]
LGTM, great job!

**Pull request merged**